File: 0897-increasing-order-search-tree/0897-increasing-order-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* cur = NULL;
    void solve(TreeNode* root) {
        if(!root) return;
        solve(root->left);
        cur->right = new TreeNode(root->val);
        cur = cur->right;
        solve(root->right);
    }
    TreeNode* increasingBST(TreeNode* root) {
        TreeNode* dummy = new TreeNode(0);
        cur = dummy;
        solve(root);
        return dummy->right;
    }
};

File: 0897-increasing-order-search-tree/0897-increasing-order-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode cur;
    void solve(TreeNode root) {
        if(root == null) return;
        solve(root.left);
        cur.right = new TreeNode(root.val);
        cur = cur.right;
        solve(root.right);
    }
    public TreeNode increasingBST(TreeNode root) {
        TreeNode dummy = new TreeNode(0);
        cur = dummy;
        solve(root);
        return dummy.right;
    }
}